rust_src/src/syntax.rs

```diff
@@ -104,4 +109,88 @@ pub fn copy_syntax_table(mut table: LispObject) -> LispObject {
     copy
 }
 
+/// Move point forward ARG words (backward if ARG is negative).
+/// If ARG is omitted or nil, move point forward one word.
+/// Normally returns t.
+/// If an edge of the buffer or a field boundary is reached, point is
+/// left there and the function returns nil.  Field boundaries are not
+/// noticed if `inhibit-field-text-motion' is non-nil.
+///
+/// The word boundaries are normally determined by the buffer's syntax
+/// table, but `find-word-boundary-function-table', such as set up
+/// by `subword-mode', can change that.  If a Lisp program needs to
+/// move by words determined strictly by the syntax table, it should
+/// use `forward-word-strictly' instead.
+#[lisp_fn(min = "0", intspec = "^p")]
+pub fn forward_word(arg: Option<EmacsInt>) -> bool {
+    let arg = arg.unwrap_or(1);
+    let cur_buf = ThreadState::current_buffer();
+    let point = cur_buf.pt;
+
```

Comment:
While I like how this reads, it does not match the original C's intent.

Both orig_val and val are created by the call to `scan_words`. However, orig_val is not changed if it is 0 and val is.
The correct code written in this style would be (I used an assignment to simplify the return and avoid repetition.):

```
let (val, orig_val) = match unsafe { scan_words(point, arg) } {
        0 => {
           let val = if arg > 0 { cur_buf.zv } else { cur_buf.begv };
           (val, 0)
        }
        n => (n, n),
};
```